doc/neps/nep-0040-legacy-datatype-impl.rst

```diff
@@ -0,0 +1,481 @@
+   * using ``isinstance(..., float)`` for known Python types
+     (note that these rules mean that subclasses are *currently* valid).
+   * special rule for void datatypes to coerce tuples.
+2. Promoting the current dtype with the next elements dtype using
+   ``np.promote_types``.
+3. If strings are found, they are allowed to try again (see also [gh15327]_).
+
+If ``dtype=...`` is given, this DType is used unless it is a
+*flexible dtype instance* such as ``"S0"``
+(in the DType class hierarchy, this would be represented by the string DType class).
+If it is such a *flexible dtype instance* ``PyArray_AdaptFlexibleDType`` is called
+which effectively inspects all values using DType specific logic.
+That is:
+
+* Strings will use ``str(element)`` to find the length of most elements
+* Datetime64 is capable of coercing from strings and guessing the correct unit.
+
+It seems probable that during normal discovery, the ``isinstance`` should rather
+be strict ``type(element) is desired_type`` checks.
+Further, the current ``AdaptFlexibleDType`` logic should be made available to
+user DTypes and not made a secondary step, but instead be called instead of,
+or be part of, the normal discovery.
+
+
+
+Related Work
+------------
+
+* Julia has similar split of abstract and concrete types [julia-types]_. 
+
+* In Julia promotion can occur based on abstract types. If a promoter is
+  defined, it will cast the inputs and then Julia can then retry to find
+  an implementation with the new values [julia-promotion]_.
+
+* ``xnd-project`` (https://github.com/xnd-project) with ndtypes and gumath
+
+  * The ``xnd-project`` solves a similar project, however, it does not use
+    casting during function execution, so that it does not require a promotion
+    step.
+
+
+Existing ``PyArray_Descr`` fields
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Although not extensively used outside of NumPy itself, the currently
+``PyArray_Descr`` is a public structure.
+This is especially also true for the ``ArrFunctions`` structure stored in
+the ``f`` field. 
+Due to compatibility they may need remain supported for a very long time,
+with the possibility of replacing them by functions that dispatch to a newer API.
+
+The access to these structures will have to be deprecated.
+
+
+Open Issues
+-----------
+
+``np.load`` (and others) currently translate all extension dtypes to void dtypes.
+This means they cannot be stored using the ``npy`` format.
+If we wish to add support for ``npy`` format in user datatypes, a new protocol
+will have to be designed to achieve this.
+
+The additional existence of masked arrays and especially masked datatypes
+within Pandas has the interesting implications of interoperability.
+Even though NumPy itself may never have true support for masked values,
+the datatype system could grow it.
+This means support for datatypes where the actual mask information is not
+stored as part of the datatype, but as part of a bit array in the array object.
```

Comment:
This seems to be a special case combining:
* having field-clustered storage for structured dtypes, rather than element-clustered (struct of arrays vs array of structs)
* support for bit-strides

Both of these would require fairly fundamental changes to the semantics of `arr.strides` and `arr.view`, which sounds like a can of worms best left unopened.

Reply:
Ah, to clarify that. I have no intention to even think about opening the can of worms for numpy. I noted it, because pandas has/will have such types, and if it is much more straight forward, we could think about helping out there.